File: scripts/vscripts/ViSDK/ViGameCommon/Event/ViCallback.ts
import { ViDebuger } from "../../ViSystem/ViDebuger";
import { ViDoubleLink, ViDoubleLinkNode } from "../../ViStruct/ViDoubleLink";
import { ViRealTimerInstance } from "../Time/ViTimerInstance";

export abstract class ViAsynDelegateInterface
{
    private static readonly CACHE_Update_ExecList = new ViDoubleLink<ViAsynDelegateInterface>();
    public static Update(): void
    {
        if (ViAsynDelegateInterface._immediateList.NotEmpty)
        {
            let execList = ViAsynDelegateInterface.CACHE_Update_ExecList;
            //
            execList.PushListBack(ViAsynDelegateInterface._immediateList);
            //
            while (execList.NotEmpty)
            {
                let asynCallback = execList.GetHead().Data;
                //ViDebuger.AssertError(asynCallback != null);
                asynCallback._attachNode.DetachEx(null);
                asynCallback._AsynExec();
            }
            ViDebuger.AssertWarning(execList.Empty);
        }
        //
        if (ViAsynDelegateInterface._timeList.NotEmpty)
        {
            let execList = ViAsynDelegateInterface.CACHE_Update_ExecList;
            //
            let time = ViRealTimerInstance.Time;
            for (let _timeList = ViAsynDelegateInterface._timeList, iter = _timeList.GetHead(); !_timeList.IsEnd(iter);)
            {
                let iterCallback = iter.Data;
                iter = _timeList.Next(iter);
                //
                if (time >= iterCallback._execTime)
                {
                    execList.PushBack(iterCallback._attachNode);
                }
            }
            //
            while (execList.NotEmpty)
            {
                let asynCallback = execList.GetHead().Data;
                //ViDebuger.AssertError(asynCallback != null);
                asynCallback._attachNode.DetachEx(null);
                asynCallback._AsynExec();
            }
            ViDebuger.AssertWarning(execList.Empty);
        }
    }
    private static readonly _immediateList = new ViDoubleLink<ViAsynDelegateInterface>();
    private static readonly _timeList = new ViDoubleLink<ViAsynDelegateInterface>();
    //
    public TimeSpan: number = 0;
    //
    protected _AttachAsyn(): void
    {
        if (this._attachNode.IsAttach())
        {
            return;
        }
        //
        if (this.TimeSpan > 0)
        {
            this._execTime = ViRealTimerInstance.Time + this.TimeSpan;
            ViAsynDelegateInterface._timeList.PushBackEx(this._attachNode, this);
        }
        else
        {
            ViAsynDelegateInterface._immediateList.PushBackEx(this._attachNode, this);
        }
    }
    //
    abstract _AsynExec(): void;
    private _execTime: number = 0;
    protected readonly _attachNode = new ViDoubleLinkNode<ViAsynDelegateInterface>();
}

File: scripts/vscripts/Game/ViSDK/ViEvent/ViCallback.lua
local ViAsynDelegateInterface = BaseClass("ViAsynDelegateInterface")

local function ViAsynDelegateInterfaceCtor(self)
    self.TimeSpan = 0
    self._execTime = 0
    self._attachNode = ViDoubleLinkNode2.New()
end

local function Update()
    if ViAsynDelegateInterface._immediateList:IsNotEmpty() then
        local execList = ViAsynDelegateInterface.CACHE_Update_ExecList
        execList:PushBackList(ViAsynDelegateInterface._immediateList)
        while execList:IsNotEmpty() do
            local asynCallback = execList:GetHead().Data
            asynCallback._attachNode:DetachEx(nil)
            asynCallback:_AsynExec()
        end
    end
    if ViAsynDelegateInterface._timeList:IsNotEmpty() then
        local execList = ViAsynDelegateInterface.CACHE_Update_ExecList
        local time = ViRealTimerInstance.Time
        local _timeList = ViAsynDelegateInterface._timeList
        local iter = _timeList:GetHead()
        while not _timeList:IsEnd(iter) do
            local iterCallback = iter.Data
            iter = _timeList:Next(iter)
            if time > iterCallback._execTime then
                execList:PushBackNode(iterCallback._attachNode)
            end
        end
        while execList:IsNotEmpty() do
            local asynCallback = execList:GetHead().Data
            asynCallback._attachNode:DetachEx(nil)
            asynCallback:_AsynExec()
        end
    end
end

local function _AttachAsyn(self)
    if self._attachNode:IsAttach() then
        return
    end
    if self.TimeSpan > 0 then
        self._execTime = ViRealTimerInstance.Time + this.TimeSpan
        ViAsynDelegateInterface._timeList:PushBackEx(self._attachNode, self)
    else
        ViAsynDelegateInterface._immediateList:PushBackEx(self._attachNode, self);
    end
end

ViAsynDelegateInterface.__init = ViAsynDelegateInterfaceCtor
ViAsynDelegateInterface.CACHE_Update_ExecList = ViDoubleLink2.New()
ViAsynDelegateInterface._immediateList = ViDoubleLink2.New()
ViAsynDelegateInterface._timeList = ViDoubleLink2.New()

ViAsynDelegateInterface.Update = Update
ViAsynDelegateInterface._AttachAsyn = _AttachAsyn

return ViAsynDelegateInterface